mri/reconstructors/calibrationless.py

```diff
@@ -0,0 +1,108 @@
+# -*- coding: utf-8 -*-
+##########################################################################
+# pySAP - Copyright (C) CEA, 2017 - 2018
+# Distributed under the terms of the CeCILL-B license, as published by
+# the CEA-CNRS-INRIA. Refer to the LICENSE file or to
+# http://www.cecill.info/licences/Licence_CeCILL-B_V1-en.html
+# for details.
+##########################################################################
+
+"""
+This implements calibrationless reconstruction with different proximities
+"""
+# System import
+import warnings
+
+from .base import ReconstructorBase
+from ..operators import GradAnalysis, GradSynthesis, WaveletN
+
+# Third party import
+from modopt.opt.proximity import SparseThreshold
+from modopt.opt.linear import Identity
+
+
+class SparseCalibrationlessReconstructor(ReconstructorBase):
+    """ This class implements a calibrationless reconstruction based on the
+    L1-norm regularization.
+    For the Analysis case finds the solution for x of:
+        (1/2) * sum(||F x_l - yl||^2_2, n_coils) +
+                    mu * sum(||Wt x_l||_1, n_coils)
+    For the Synthesis case finds the solution of:
+        (1/2) * sum(||F Wt alpha_l - yl||^2_2, n_coils) +
+                    mu * sum(||alpha_l||_1, n_coils)
+    Parameters
+    ----------
+    fourier_op: object of class FFT, NonCartesianFFT or Stacked3DNFFT in
+                mri.operators
+        Defines the fourier operator F in the above equation.
+    linear_op: object, (optional, default None)
+        Defines the linear sparsifying operator Wt. This must operate on x and
+        have 2 functions, op(x) and adj_op(coeff) which implements the
+        operator and adjoint operator. For wavelets, this can be object of
+        class WaveletN or WaveletUD2 from mri.operators .
+        If None, sym8 wavelet with nb_scales=3 is chosen.
+    prox_op: operator, (optional default None)
+        Defines the proximity operator for the regularization function.
+        For example, for L1 Norm, the proximity operator is Thresholding
+        If None, the proximity opertaor is defined as Soft thresholding
+        of wavelet coefficients with mu value as specified.
+    mu: float or np.ndarray, (optional, default 0)
+        If prox_op is None, the value of mu is used to form a proximity
+        operator that is soft thresholding of the wavelet coefficients.
+        If prox_op is specified, this is ignored.
+    lips_calc_max_iter: int, default 10
```

Comment:
`gradient _method` is missing also, please consider changing it to `gradient_formulation`